### PR TITLE
add waveglow

### PR DESCRIPTION
**Pull request by upvenly**

### CODE REVIEW

File: training/nvidia/WaveGlow-pytorch/README.md
Comment (yuzhou03) [inline training/nvidia/WaveGlow-pytorch/README.md:1]: 厂商目录下，和config同级的，缺少一个extern目录。

File: training/benchmarks/WaveGlow/pytorch/config/mutable_params.py
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/config/mutable_params.py:1]: add amp.
或者在base中把amp设置为false
厂商有不支持amp的，不能强制他们为True

File: training/benchmarks/WaveGlow/pytorch/dataloaders/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        mask = mask.any(dim=2)
+        masks.append(mask)
+    if masks:
+        masks = torch.stack(masks, dim=0)
+    else:
+        masks = torch.zeros((0, height, width), dtype=torch.uint8)
+    return masks
+
+
+class ConvertCocoPolysToMask(object):
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/dataloaders/dataloader.py:104]: 这几个是干啥的？看名字是CV领域的func，和waveglow完全没关系

File: training/benchmarks/WaveGlow/pytorch/loss/loss_function.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def get_loss_function(loss_function, sigma=1.0):
+    if loss_function == 'WaveGlow':
+        loss = WaveGlowLoss(sigma=sigma)
+    else:
+        raise NotImplementedError(
+            "unknown loss function requested: {}".format(loss_function))
+
+    loss.cuda()
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/loss/loss_function.py:61]: 这个需要联合前面的确认下，Loss输入的output和clean audio是在GPUdevice上吗？是的话这个return出来的loss应该也在？

File: training/benchmarks/WaveGlow/pytorch/model/model.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+# *****************************************************************************
+import torch
+
+torch._C._jit_set_autocast_mode(False)
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+
+
+@torch.jit.script
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/model/model.py:35]: 虽然这个是copied但也要确认下，inference用的jit库，厂商不一定支持。jit底层实现方式是否与torch.cuda.amp(enable=None)相同？

File: training/benchmarks/WaveGlow/pytorch/run_pretraining.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
+# *****************************************************************************
+#  Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/run_pretraining.py:2]: runpretraining是我们自己写的，不要放nvidia的license

File: training/benchmarks/WaveGlow/pytorch/run_pretraining.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    if not config.do_train:
+        return config, training_state
+
+    trainset = build_train_dataset(config)
+    train_loader = build_train_dataloader(trainset, config)
+    valset = build_eval_dataset(config)
+    val_loader = build_eval_dataloader(valset, config)
+
+    start_epoch = [0]
+    start_epoch = start_epoch[0]
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/run_pretraining.py:101]: 这两句是干啥用的

File: training/benchmarks/WaveGlow/pytorch/train/trainer.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        self.model = create_model(self.config)
+        self.model.to(self.config.device)
+        self.model = self.adapter.convert_model(self.model)
+        self.model = self.adapter.model_to_fp16(self.model, self.config)
+        self.model = self.adapter.model_to_ddp(self.model, self.config)
+        self.optimizer = create_optimizer(self.model, self.config)
+        self.scaler = self.adapter.create_grad_scaler(self.config)
+        self.criterion = create_criterion(self.config)
+        self.batch_to_gpu = data_function.get_batch_to_gpu(self.config.name)
+        self.world_size = dist_pytorch.get_world_size()
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/train/trainer.py:44]: 是否需要在trainer里面存这个？

File: training/benchmarks/WaveGlow/pytorch/train/trainer.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        epoch_start_time = time.perf_counter()
+        # used to calculate avg items/sec over epoch
+        reduced_num_items_epoch = 0
+
+        train_epoch_items_per_sec = 0.0
+        num_iters = 0
+        reduced_loss = 0
+
+        self.model.train()
+        if self.world_size > 1:
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/train/trainer.py:59]: 是不是只有这里用了worldsize？可以改成config里其他的item，如config.distributed

File: training/benchmarks/faster_rcnn/pytorch/train/trainer.py
Comment (shh2000) [inline training/benchmarks/faster_rcnn/pytorch/train/trainer.py:1]: 这两个修改是为了什么

File: training/nvidia/WaveGlow-pytorch/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+### 数据集下载
+
+[数据集下载](../../benchmarks/tacotron2/README.md#数据集下载地址)
```
Comment (shh2000) [inline training/nvidia/WaveGlow-pytorch/README.md:3]: 是否改成waveglow，还是共用tacotron的

File: training/benchmarks/WaveGlow/pytorch/model/models.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+def get_model(model_name,
+              model_config,
+              cpu_run,
+              uniform_initialize_bn_weight=False,
+              forward_is_infer=False,
+              jittable=False):
+    """ Code chooses a model based on name"""
+    model = None
+    if model_name == 'Tacotron2':
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/model/models.py:65]: rm redundent

File: training/benchmarks/WaveGlow/pytorch/model/models.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            model.forward = model.infer
+
+    else:
+        raise NotImplementedError(model_name)
+
+    if uniform_initialize_bn_weight:
+        init_bn(model)
+
+    if not cpu_run:
+        model = model.cuda()
```
Comment (shh2000) [inline training/benchmarks/WaveGlow/pytorch/model/models.py:89]: trainer中已有to_device。两次重复的to_device是否会导致错误？已经to(gpu 3)的model，调用cuda是否有风险被转到gpu 0？

File: training/benchmarks/WaveGlow/pytorch/utils/utils.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    rt = tensor.clone()
+    dist.all_reduce(rt, op=dist.ReduceOp.SUM)
+    if rt.is_floating_point():
+        rt = rt / num_gpus
+    else:
+        rt = torch.div(rt, num_gpus, rounding_mode='floor')
+    return rt
+
+
+def init_dllogger(config):
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/utils/utils.py:128]: DLLogger是必须的吗？

File: training/benchmarks/WaveGlow/pytorch/train/evaluator.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            val_loss = 0.0
+            num_iters = 0
+            val_items_per_sec = 0.0
+            for i, batch in enumerate(val_loader):
+                torch.cuda.synchronize()
+                iter_start_time = time.perf_counter()
+
+                x, y, num_items = batch_to_gpu(batch)
+                #AMP upstream autocast
+                with torch.cuda.amp.autocast(enabled=amp_run):
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/train/evaluator.py:27]: 可能厂商的加速卡不支持torch.cuda.amp.autocast算子，放到training_adapater里，可以被vendor覆盖

File: training/benchmarks/WaveGlow/pytorch/train/trainer.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+    def __init__(self, training_state, config):
+        super(Trainer, self).__init__()
+        self.adapter = trainer_adapter
+        self.training_state = training_state
+        self.config = config
+        self.logger = init_dllogger(config)
+
+    def init(self):
+        # torch.set_num_threads(1)
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/train/trainer.py:33]: 注释语句无需提交

File: training/benchmarks/WaveGlow/pytorch/train/trainer.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            self.model.zero_grad()
+
+            x, y, num_items = self.batch_to_gpu(batch)
+
+            #AMP upstream autocast
+            with torch.cuda.amp.autocast(enabled=True):
+                y_pred = self.model(x)
+                loss = self.criterion(y_pred, y)
+
+            if self.world_size > 1:
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/train/trainer.py:86]: 可以用config.distributed来判断

File: training/benchmarks/WaveGlow/pytorch/train/trainer_adapter.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    # To prevent OOM for model sizes that cannot fit in GPU memory in full precision
+    if config.fp16:
+        main_proc_print(" > use fp16...")
+        model.half()
+    return model
+
+
+def model_to_ddp(model, config):
+    """model_to_ddp"""
+    if dist.is_available() and dist.is_initialized():
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/train/trainer_adapter.py:29]: 可以复用benchmarks/driver/dist_pytorch.py下的is_dist_avail_and_initialized函数

File: training/benchmarks/WaveGlow/pytorch/config/_base.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+tar_val_loss = -5.72  #https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechSynthesis/Tacotron2
+
+save_checkpoint = True
+
+#Perf
+do_train = True
+local_rank = -1
+log_freq = 1
+output = "training/result/"
+log_file = "nvlog.json"
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/config/_base.py:14]: 这是原始repo的args参数，Perf里不用保存为nvlog.json

File: training/nvidia/WaveGlow-pytorch/README.md
Comment (upvenly) [reply training/nvidia/WaveGlow-pytorch/README.md:1]: 不需要extern也需要这个目录吗

File: training/benchmarks/WaveGlow/pytorch/utils/utils.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+                         'learning_rate changed':
+                         str(optimizer.param_groups[0]['lr']) + " -> " +
+                         str(lr)
+                     })
+
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+
+
+def save_checkpoint(model, optimizer, scaler, epoch, config, output_dir,
```
[the remark this answers is not in the document]
Comment (upvenly) [reply training/benchmarks/WaveGlow/pytorch/utils/utils.py:35]: 有需要，最后版本关闭。以及是否可以不要再提这种非必要的review意见。

File: training/benchmarks/WaveGlow/pytorch/dataloaders/dataloader.py (hunk 1 above)
Comment (upvenly) [reply training/benchmarks/WaveGlow/pytorch/dataloaders/dataloader.py:104]: 冗余代码，已删除

File: training/nvidia/WaveGlow-pytorch/README.md
Comment (yuzhou03) [reply training/nvidia/WaveGlow-pytorch/README.md:1]: 是的，没有extern目录启动run.py会报错。

File: training/benchmarks/WaveGlow/pytorch/utils/utils.py (hunk 11 above)
Comment (upvenly) [reply training/benchmarks/WaveGlow/pytorch/utils/utils.py:128]: 是的

**Comment by upvenly** [comment]
以下文件参考源代码，不需要review：
training/benchmarks/WaveGlow/pytorch/model/models.py
training/benchmarks/WaveGlow/pytorch/model/model.py
training/benchmarks/WaveGlow/pytorch/model/model_parser.py
training/benchmarks/WaveGlow/pytorch/dataloaders/data_function.py
training/benchmarks/WaveGlow/pytorch/filelists
training/benchmarks/WaveGlow/pytorch/loss/loss_function.py
training/benchmarks/WaveGlow/pytorch/tacotron2_common
training/benchmarks/WaveGlow/pytorch/utils

File: training/benchmarks/WaveGlow/pytorch/README.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
+### 模型信息
+- Introduction
+
+The WaveGlow model is a flow-based generative model that generates audio samples from Gaussian distribution using mel-spectrogram conditioning (Figure 2). During training, the model learns to transform the dataset distribution into spherical Gaussian distribution through a series of flows. One step of a flow consists of an invertible convolution, followed by a modified WaveNet architecture that serves as an affine coupling layer. During inference, the network is inverted and audio samples are generated from the Gaussian distribution. Our implementation uses 512 residual channels in the coupling layer.
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/README.md:4]: 注意代词, we/us/our

File: training/benchmarks/WaveGlow/pytorch/schedulers/__init__.py [hunk 19]
```diff
@@ -0,0 +1,8 @@
+from torch.optim.lr_scheduler import MultiStepLR
+
+
+def create_scheduler(optimizer, args):
+    lr_scheduler = MultiStepLR(optimizer,
+                               milestones=args.lr_steps,
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/schedulers/__init__.py:6]: 这里引用了lr_steps和lr_gamma，_base.py中没有配置？

File: training/benchmarks/faster_rcnn/pytorch/train/trainer.py
Comment (yuzhou03) [inline training/benchmarks/faster_rcnn/pytorch/train/trainer.py:1]: 无需修改faster_rcnn的代码

File: training/nvidia/WaveGlow-pytorch/README.md [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+   - 加速卡驱动版本：470.129.06
+   - Docker 版本：20.10.16
+   - 训练框架版本：pytorch-1.13
+   - 依赖软件版本：无
+
+
+### 运行情况
+| 训练资源 | 配置文件        | 运行时长(s) | 目标val_loss | 收敛val_loss | 性能(samples/s) |
+| -------- | --------------- | ----------- | ------------ | ------------ | --------------- |
+| 单机8卡  | config_A100x1x8 |     |      |        |           |
```
Comment (yuzhou03) [inline training/nvidia/WaveGlow-pytorch/README.md:24]: 补上运行数据

File: training/benchmarks/WaveGlow/pytorch/train/trainer_adapter.py [hunk 21]
```diff
@@ -0,0 +1,41 @@
+import os
+import sys
+
+import torch
+import torch.distributed as dist
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/train/trainer_adapter.py:5]: dist在本文件中未使用

File: training/benchmarks/driver/helper.py [hunk 22]
```diff
@@ -17,18 +17,19 @@ class InitHelper:
...
         config = self.config
         model_driver = Driver(config, config.mutable_params)
-        model_driver.setup_config(argparse.ArgumentParser(config.name))
+        parser = argparse.ArgumentParser(config.name) if parser is None else parser
+        model_driver.setup_config(parser)
         model_driver.setup_modules(global_module, local_module)
         check.check_config(model_driver.config)
         self.update_local_rank()
-        return model_driver
+        return model_driver, parser
```
Comment (yuzhou03) [inline training/benchmarks/driver/helper.py:32]: 注意：这里增加了一个返回参数，新添加的模型的run_pretraining.py中调用的地方都要修改，接受或忽略第二个返回值。不然会报错。

File: training/nvidia/docker_image/pytorch_1.8/pytorch_install.sh
Comment (yuzhou03) [inline training/nvidia/docker_image/pytorch_1.8/pytorch_install.sh:1]: 文件内容为空，有必要提交吗

File: training/benchmarks/WaveGlow/pytorch/loss/__init__.py [hunk 23]
```diff
@@ -0,0 +1,10 @@
+from . import loss_function
+
+
+def create_criterion(args):
+    try:
+        sigma = args.sigma
```
Comment (yuzhou03) [inline training/benchmarks/WaveGlow/pytorch/loss/__init__.py:6]: _base.py中没有配置sigma, 运行没报错？

**Review by yuzhou03 — APPROVED** [review]
